File: sdk-api/src/test/java/dev/restate/sdk/CallTest.java
// Copyright (c) 2023 - Restate Software, Inc., Restate GmbH
//
// This file is part of the Restate Java SDK,
// which is released under the MIT license.
//
// You can find a copy of the license in file LICENSE in the root
// directory of this repository or package, or at
// https://github.com/restatedev/sdk-java/blob/main/LICENSE
package dev.restate.sdk;

import static dev.restate.sdk.JavaBlockingTests.testDefinitionForService;

import dev.restate.common.Slice;
import dev.restate.common.Target;
import dev.restate.sdk.core.CallTestSuite;
import dev.restate.sdk.core.TestDefinitions.TestInvocationBuilder;
import dev.restate.serde.Serde;
import java.util.Map;

public class CallTest extends CallTestSuite {

  @Override
  protected TestInvocationBuilder oneWayCall(
      Target target, String idempotencyKey, Map<String, String> headers, Slice body) {
    return testDefinitionForService(
        "OneWayCall",
        Serde.VOID,
        Serde.VOID,
        (context, unused) -> {
          context.send(
              target,
              body.toByteArray(),
              SendOptions.builder().headers(headers).idempotencyKey(idempotencyKey).build());
          return null;
        });
  }

  @Override
  protected TestInvocationBuilder implicitCancellation(Target target, Slice body) {
    return testDefinitionForService(
        "Implicit cancellation",
        Serde.VOID,
        Serde.RAW,
        (context, unused) -> context.call(target, body.toByteArray()).await());
  }
}

File: sdk-api-kotlin/src/test/kotlin/dev/restate/sdk/kotlin/CallTest.kt
// Copyright (c) 2023 - Restate Software, Inc., Restate GmbH
//
// This file is part of the Restate Java SDK,
// which is released under the MIT license.
//
// You can find a copy of the license in file LICENSE in the root
// directory of this repository or package, or at
// https://github.com/restatedev/sdk-java/blob/main/LICENSE
package dev.restate.sdk.kotlin

import dev.restate.common.Slice
import dev.restate.common.Target
import dev.restate.sdk.core.CallTestSuite
import dev.restate.sdk.kotlin.KotlinCoroutinesTests.Companion.testDefinitionForService
import dev.restate.serde.Serde

class CallTest : CallTestSuite() {

  override fun oneWayCall(
      target: Target,
      idempotencyKey: String,
      headers: Map<String, String>,
      body: Slice
  ) =
      testDefinitionForService("OneWayCall") { ctx, _: Unit ->
        ctx.send(
            target,
            Serde.SLICE,
            body,
            sendOptions {
              this.idempotencyKey = idempotencyKey
              this.headers = LinkedHashMap(headers)
            })
      }

  override fun implicitCancellation(target: Target, body: Slice) =
      testDefinitionForService("ImplicitCancellation") { ctx, _: Unit ->
        ctx.call(target, Serde.SLICE, Serde.RAW, body)
      }
}
